File: AccelerationTest/tools/dtn_config.py
"""
DTN (LRU 0x2600) configuration encoder.

Wire format, reverse-engineered from and cross-checked against three independent
implementations in this repository:

  * RemoteConfigSender/main.cpp        - the reference config blobs
  * dpdk/src/HealthMonitor/HealthMonitor.c - the 0x52 status query
  * FirmwareUpdater/fpga_firmware_loader.py - LRU_ID / OpType / trailing seq byte

  Ethernet   dst 03:00:00:00:<VL hi>:<VL lo>   src 02:00:00:00:00:<net>
             ethertype 0x0800, or 0x8100 + TCI + 0x0800 when tagged
  IPv4       src 10.1.33.1  dst 224.224.<VL hi>.<VL lo>  TTL 1  proto UDP
  UDP        100 -> 100, checksum disabled
  Payload    LRU_ID(2) OpType(1) | Addr(1) Len(2) Data(Len) | M | Addr Len Data | ...
  Trailer    1-byte AFDX sequence, OUTSIDE the IP total_length

`M` is a one-byte end-of-block marker; see `block_marker`.
"""

import struct

LRU_DTN      = 0x2600
OP_WRITE     = 0x57
OP_READ      = 0x52

NET_A        = 0x20
NET_B        = 0x40

SRC_IP       = bytes([10, 1, 33, 1])
UDP_PORT     = 100
MAX_FRAME    = 1518

# Config-space start addresses (the "CfgType" byte)
ADDR_ES_GLOBAL   = 0x10   # end-system globals (VL count, Lmax)
ADDR_ES_PARAMS   = 0x17   # end-system parameters
ADDR_PTP         = 0x46   # PTP session table
ADDR_SW_BEGIN    = 0x70   # switch table: begin
ADDR_SW_END      = 0x71   # switch table: end
ADDR_VL_TABLE    = 0x72   # switch table: VL records (14 bytes each)
ADDR_PORT_TABLE  = 0x73   # per-port table (4 bytes each)
ADDR_SW_MISC     = 0x74   # switch misc

VL_RECORD_LEN    = 14
DTN_PORT_COUNT   = 35     # ports 0..34 (34 = CPU)

# Largest record count that keeps an untagged frame within 1518 bytes:
#   14 eth + 20 ip + 8 udp + 3 hdr + 3 block + 14*N + 1 seq <= 1518
MAX_RECORDS_PER_BLOCK = 104


def block_marker(addr: int) -> int:
    """One-byte end-of-block marker that follows a block's data.

    Derived empirically; reproduces every marker in the reference blobs:
      0x70..0x7F  ->  (addr & 0x0F) | 0x80      (0x70->0x80, 0x72->0x82, ...)
      otherwise   ->  addr + 2                  (0x10->0x12, 0x17->0x19, 0x46->0x48)
    """
    if 0x70 <= addr <= 0x7F:
        return (addr & 0x0F) | 0x80
    return (addr + 2) & 0xFF


# ---------------------------------------------------------------------------
# VL record
# ---------------------------------------------------------------------------

class VlRecord:
    """One row of the switch VL table - the binary form of an XML <VL .../> line.

      off  size  XML attribute
      [0]   2    ID
      [2]   2    BAG / PRIORITY word          (0x0602 for BAG=1MS PRIORITY=LOW)
      [4]   1    JITTER, milliseconds
      [5]   1    LMIN
      [6]   2    flag nibble | LMAX (12 bit)  (0x9 = ENABLE, LOW priority)
      [8]   1    DESTPORT bits 34..32
      [9]   1    SRCPORT
      [10]  4    DESTPORT bits 31..0
    """

    def __init__(self, vl_id, src_port, dest_ports,
                 lmax=1518, lmin=64, jitter_ms=0,
                 bag_word=0x0602, flags=0x9):
        self.vl_id     = vl_id
        self.src_port  = src_port
        self.dest      = frozenset(dest_ports)
        self.lmax      = lmax
        self.lmin      = lmin
        self.jitter_ms = jitter_ms
        self.bag_word  = bag_word
        self.flags     = flags

    def dest_mask(self) -> int:
        m = 0
        for p in self.dest:
            if not 0 <= p < DTN_PORT_COUNT:
                raise ValueError(f"VL {self.vl_id}: destination port {p} out of range 0..34")
            m |= 1 << p
        return m

    def encode(self) -> bytes:
        if not 0 <= self.src_port < DTN_PORT_COUNT:
            raise ValueError(f"VL {self.vl_id}: source port {self.src_port} out of range 0..34")
        if not 0 < self.lmax <= 0xFFF:
            raise ValueError(f"VL {self.vl_id}: LMAX {self.lmax} does not fit in 12 bits")
        mask = self.dest_mask()
        return struct.pack(
            ">HHBBHBBI",
            self.vl_id,
            self.bag_word,
            self.jitter_ms,
            self.lmin,
            ((self.flags & 0xF) << 12) | (self.lmax & 0xFFF),
            (mask >> 32) & 0xFF,
            self.src_port,
            mask & 0xFFFFFFFF,
        )

    @staticmethod
    def decode(raw: bytes) -> "VlRecord":
        (vl, bag, jit, lmin, fl_lmax, dhi, src, dlo) = struct.unpack(">HHBBHBBI", raw)
        mask = (dhi << 32) | dlo
        return VlRecord(vl, src, {p for p in range(DTN_PORT_COUNT) if mask >> p & 1},
                        lmax=fl_lmax & 0xFFF, lmin=lmin, jitter_ms=jit,
                        bag_word=bag, flags=fl_lmax >> 12)

    def __repr__(self):
        d = sorted(self.dest)
        return (f"VL {self.vl_id:5d}  src={self.src_port:2d} -> dst={d}  "
                f"LMIN={self.lmin} LMAX={self.lmax} JITTER={self.jitter_ms}MS "
                f"bag=0x{self.bag_word:04x} flags=0x{self.flags:x}")


def encode_port_table(value=0x0112, port_count=34) -> bytes:
    """Per-port table (address 0x73): port_id(2 BE) + 2-byte value, one row per port."""
    return b"".join(struct.pack(">HH", p, value) for p in range(port_count))


# ---------------------------------------------------------------------------
# Datagram / frame assembly
# ---------------------------------------------------------------------------

def build_payload(blocks, op=OP_WRITE, lru=LRU_DTN, terminate=False) -> bytes:
    """blocks: list of (addr, data). A marker separates consecutive blocks;
    `terminate` appends the final block's marker (last datagram of a group)."""
    out = bytearray(struct.pack(">HB", lru, op))
    for i, (addr, data) in enumerate(blocks):
        if i:
            out.append(block_marker(blocks[i - 1][0]))
        out += struct.pack(">BH", addr, len(data)) + data
    if terminate:
        out.append(block_marker(blocks[-1][0]))
    return bytes(out)


def ip_checksum(header: bytes) -> int:
    total = 0
    for i in range(0, len(header), 2):
        total += (header[i] << 8) | header[i + 1]
    while total >> 16:
        total = (total & 0xFFFF) + (total >> 16)
    return ~total & 0xFFFF


def build_frame(payload: bytes, seq: int, vl_id=0, vlan=None, net=NET_A) -> bytes:
    vl = struct.pack(">H", vl_id)
    eth = b"\x03\x00\x00\x00" + vl + b"\x02\x00\x00\x00\x00" + bytes([net])
    eth += (b"\x81\x00" + struct.pack(">H", vlan & 0xFFF) if vlan is not None else b"") + b"\x08\x00"

    udp = struct.pack(">HHHH", UDP_PORT, UDP_PORT, 8 + len(payload), 0)
    ip_len = 20 + len(udp) + len(payload)
    ip = (b"\x45\x00" + struct.pack(">H", ip_len) + b"\xd4\x3b\x00\x00\x01\x11"
          + b"\x00\x00" + SRC_IP + b"\xe0\xe0" + vl)
    ip = ip[:10] + struct.pack(">H", ip_checksum(ip)) + ip[12:]

    frame = eth + ip + udp + payload + bytes([seq & 0xFF])
    if len(frame) > MAX_FRAME:
        raise ValueError(f"frame is {len(frame)} bytes, over the {MAX_FRAME} limit")
    return frame


def next_seq(seq: int) -> int:
    """The device's 1-byte AFDX counter: 0 is only ever the first value, then 1..255."""
    return 1 if seq >= 255 else seq + 1


def chunk_vl_records(records, per_block=MAX_RECORDS_PER_BLOCK):
    for i in range(0, len(records), per_block):
        yield records[i:i + per_block]

File: DtnConfigurator/src/DtnConfig.c
#include "DtnConfig.h"

#include <string.h>

#define ETH_HDR_LEN   14
#define VLAN_TAG_LEN  4
#define IP_HDR_LEN    20
#define UDP_HDR_LEN   8
#define DTN_UDP_PORT  100

static void put_be16(uint8_t *p, uint16_t v)
{
    p[0] = (uint8_t)(v >> 8);
    p[1] = (uint8_t)v;
}

static void put_be32(uint8_t *p, uint32_t v)
{
    p[0] = (uint8_t)(v >> 24);
    p[1] = (uint8_t)(v >> 16);
    p[2] = (uint8_t)(v >> 8);
    p[3] = (uint8_t)v;
}

static uint16_t get_be16(const uint8_t *p)
{
    return (uint16_t)((p[0] << 8) | p[1]);
}

static uint32_t get_be32(const uint8_t *p)
{
    return ((uint32_t)p[0] << 24) | ((uint32_t)p[1] << 16) |
           ((uint32_t)p[2] << 8) | p[3];
}

static uint16_t ip_checksum(const uint8_t *header, size_t len)
{
    uint32_t sum = 0;
    for (size_t i = 0; i + 1 < len; i += 2)
        sum += (uint32_t)((header[i] << 8) | header[i + 1]);
    while (sum >> 16)
        sum = (sum & 0xFFFF) + (sum >> 16);
    return (uint16_t)~sum;
}

void dtn_vl_init(dtn_vl_t *vl, uint16_t vl_id, uint8_t src_port, uint64_t dest_mask)
{
    vl->vl_id     = vl_id;
    vl->src_port  = src_port;
    vl->dest_mask = dest_mask;
    vl->lmax      = 1518;
    vl->lmin      = 64;
    vl->jitter_ms = 0;
    vl->bag_word  = DTN_BAG_WORD_1MS;
    vl->flags     = DTN_VL_FLAG_ENABLE | DTN_VL_FLAG_RESERVED;
}

int dtn_vl_encode(const dtn_vl_t *vl, uint8_t out[DTN_VL_RECORD_LEN])
{
    if (vl->src_port >= DTN_PORT_COUNT)
        return -1;
    if (vl->lmax == 0 || vl->lmax > 0x0FFF)
        return -1;
    if (vl->dest_mask >> DTN_PORT_COUNT)
        return -1;

    put_be16(out + 0, vl->vl_id);
    put_be16(out + 2, vl->bag_word);
    out[4] = vl->jitter_ms;
    out[5] = vl->lmin;
    put_be16(out + 6, (uint16_t)(((vl->flags & 0x0F) << 12) | (vl->lmax & 0x0FFF)));
    out[8] = (uint8_t)((vl->dest_mask >> 32) & 0xFF);
    out[9] = vl->src_port;
    put_be32(out + 10, (uint32_t)(vl->dest_mask & 0xFFFFFFFFu));
    return 0;
}

void dtn_vl_decode(const uint8_t in[DTN_VL_RECORD_LEN], dtn_vl_t *vl)
{
    uint16_t flags_lmax = get_be16(in + 6);

    vl->vl_id     = get_be16(in + 0);
    vl->bag_word  = get_be16(in + 2);
    vl->jitter_ms = in[4];
    vl->lmin      = in[5];
    vl->flags     = (uint8_t)(flags_lmax >> 12);
    vl->lmax      = flags_lmax & 0x0FFF;
    vl->src_port  = in[9];
    vl->dest_mask = ((uint64_t)in[8] << 32) | get_be32(in + 10);
}

uint8_t dtn_block_marker(uint8_t addr)
{
    if (addr >= 0x70 && addr <= 0x7F)
        return (uint8_t)((addr & 0x0F) | 0x80);
    return (uint8_t)(addr + 2);
}

int dtn_build_payload(const dtn_block_t *blocks, size_t count, uint8_t op,
                      bool terminate, uint8_t *out, size_t cap)
{
    size_t n = 0;

    if (count == 0)
        return -1;

    /* LRU_ID(2) + OpType(1) */
    if (cap < 3)
        return -1;
    put_be16(out, DTN_LRU_ID);
    out[2] = op;
    n = 3;

    for (size_t i = 0; i < count; i++) {
        size_t need = (i ? 1u : 0u) + 3u + blocks[i].len;
        if (n + need > cap)
            return -1;
        if (i)
            out[n++] = dtn_block_marker(blocks[i - 1].addr);
        out[n++] = blocks[i].addr;
        put_be16(out + n, blocks[i].len);
        n += 2;
        memcpy(out + n, blocks[i].data, blocks[i].len);
        n += blocks[i].len;
    }

    if (terminate) {
        if (n + 1 > cap)
            return -1;
        out[n++] = dtn_block_marker(blocks[count - 1].addr);
    }
    return (int)n;
}

int dtn_build_frame(const uint8_t *payload, size_t len, uint8_t seq,
                    uint16_t vl_id, int vlan, uint8_t net, uint8_t *out, size_t cap)
{
    size_t tagged = (vlan >= 0) ? VLAN_TAG_LEN : 0;
    size_t total  = ETH_HDR_LEN + tagged + IP_HDR_LEN + UDP_HDR_LEN + len + 1;
    size_t n = 0;
    uint8_t *ip;

    if (total > DTN_MAX_FRAME || total > cap)
        return -1;

    /* Ethernet: the low 16 bits of the destination MAC carry the VL id. */
    out[n++] = 0x03; out[n++] = 0x00; out[n++] = 0x00; out[n++] = 0x00;
    put_be16(out + n, vl_id); n += 2;
    out[n++] = 0x02; out[n++] = 0x00; out[n++] = 0x00; out[n++] = 0x00;
    out[n++] = 0x00; out[n++] = net;
    if (vlan >= 0) {
        out[n++] = 0x81; out[n++] = 0x00;
        put_be16(out + n, (uint16_t)(vlan & 0x0FFF)); n += 2;
    }
    out[n++] = 0x08; out[n++] = 0x00;

    /* IPv4. The identification field is fixed at 0xd43b in every reference
     * frame; keep it so generated frames stay comparable with them. */
    ip = out + n;
    ip[0] = 0x45; ip[1] = 0x00;
    put_be16(ip + 2, (uint16_t)(IP_HDR_LEN + UDP_HDR_LEN + len));
    ip[4] = 0xd4; ip[5] = 0x3b;
    ip[6] = 0x00; ip[7] = 0x00;
    ip[8] = 0x01;               /* TTL */
    ip[9] = 0x11;               /* UDP */
    ip[10] = 0x00; ip[11] = 0x00;
    ip[12] = 10; ip[13] = 1; ip[14] = 33; ip[15] = 1;
    ip[16] = 0xe0; ip[17] = 0xe0;
    put_be16(ip + 18, vl_id);
    put_be16(ip + 10, ip_checksum(ip, IP_HDR_LEN));
    n += IP_HDR_LEN;

    /* UDP, checksum disabled. */
    put_be16(out + n, DTN_UDP_PORT); n += 2;
    put_be16(out + n, DTN_UDP_PORT); n += 2;
    put_be16(out + n, (uint16_t)(UDP_HDR_LEN + len)); n += 2;
    put_be16(out + n, 0); n += 2;

    memcpy(out + n, payload, len);
    n += len;

    /* AFDX sequence byte, deliberately outside the IP total_length. */
    out[n++] = seq;
    return (int)n;
}

uint8_t dtn_next_seq(uint8_t seq)
{
    return (uint8_t)(seq >= 255 ? 1 : seq + 1);
}

size_t dtn_encode_port_table(uint8_t *out, size_t cap, uint16_t value, uint8_t port_count)
{
    size_t need = (size_t)port_count * 4u;

    if (need > cap)
        return 0;
    for (uint8_t p = 0; p < port_count; p++) {
        put_be16(out + p * 4, p);
        put_be16(out + p * 4 + 2, value);
    }
    return need;
}
